File: src/数组、链表、跳表/Linked List/25.k-个一组翻转链表.cs
/*
 * @lc app=leetcode.cn id=25 lang=csharp
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseKGroup(ListNode head, int k) {
        var hair = new ListNode { next = head };
        var pre = hair;
        while(head != null) {
            var tail = pre;
            for (int i = 0; i < k; i++) {
                if(tail.next == null) {
                    return hair.next;
                }
                tail = tail.next;
            }
            var nex = tail.next;
            var res = ReverseList(head, tail);
            head = res[0];
            tail = res[1];
            pre.next = head;
            tail.next = nex;
            pre = tail;
            head = tail.next;
        }
        return hair.next;
    }
    public List<ListNode> ReverseList(ListNode head, ListNode tail) {
        if(head == tail) {
            return new List<ListNode> { head, head };
        }
        var res = ReverseList(head.next, tail);
        res[1].next = head;
        head.next = null;

        return new List<ListNode> {res[0], head};
    }
}
// @lc code=end

File: src/数组、链表、跳表/Linked List/25.k-个一组翻转链表.go
/*
 * @lc app=leetcode.cn id=25 lang=golang
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	hair := &ListNode{Next: head}
	pre := hair
	for head != nil {
		tail := pre
		for i := 0; i < k; i++ {
			tail = tail.Next
			if tail == nil {
				return hair.Next
			}
		}
		nex := tail.Next
		head, tail = reverseList(head, tail)
		pre.Next = head
		tail.Next = nex
		pre = tail
		head = tail.Next
	}
	return hair.Next
}

func reverseList(head, tail *ListNode) (*ListNode, *ListNode) {
	if(head == tail) {
		return tail, head
	}
	resHead, resTail := reverseList(head.Next, tail)
	resTail.Next = head
	head.Next = nil
	return resHead, head
}
// @lc code=end

